index.js

```diff
@@ -939,26 +938,27 @@
 
     return isEmpty(values) ?
       [Unknown] :
-      or(Z.reduce(refine, env, values), loose ? [Inconsistent] : []);
+      or(Z.reduce(refine, env, values), [Inconsistent]);
   }
 
-  //  rejectInconsistent :: Array Type -> Array Type
-  function rejectInconsistent(types) {
-    return types.filter(function(t) {
-      return t.type !== INCONSISTENT && t.type !== UNKNOWN;
-    });
+  //  isConsistent :: Type -> Boolean
+  function isConsistent(t) {
+    return t.type === UNARY   ? isConsistent(t.types.$1.type) :
+           t.type === BINARY  ? isConsistent(t.types.$1.type) &&
+                                isConsistent(t.types.$2.type) :
+           /* else */           t.type !== INCONSISTENT;
   }
 
   //  determineActualTypesStrict :: (Array Type, Array Any) -> Array Type
   function determineActualTypesStrict(env, values) {
-    var types$ = _determineActualTypes(false, env, [], values);
-    return rejectInconsistent(types$);
+    return _determineActualTypes(env, [], values)
+           .filter(isConsistent);
   }
 
   //  determineActualTypesLoose :: (Array Type, Array Any) -> Array Type
   function determineActualTypesLoose(env, values) {
-    var types$ = _determineActualTypes(true, env, [], values);
-    return rejectInconsistent(types$);
+    return _determineActualTypes(env, [], values)
+           .filter(function(t) { return t.type !== INCONSISTENT; });
   }
```

Comment:
The difference between the `*Strict` and `*Loose` functions is that the latter permits inconsistent types such as `Array ???` in order to mention these in error messages.